parflow_data_management/transport/tasks.py

```diff
@@ -37,3 +39,28 @@ def generate_key_and_passphrase(user_id, key_pair_id):
     async_to_sync(channel_layer.group_send)(
         group, {"type": "public.key.with.passphrase", "data": data}
     )
+
+
+@shared_task
+def unlock_key_pair(key_pair_id, passphrase):
+    # Retrieve encrypted private key from the database
+    key_pair = KeyPair.objects.get(pk=key_pair_id)
+    private_key_encrypted = key_pair.private_key_encrypted
+
+    # Attempt decryption
+    key_obj = RSAKey.from_private_key(
+        io.StringIO(private_key_encrypted), password=passphrase
+    )
+
+    # If it succeeded, get the unencrypted private key
+    desc = io.StringIO()
+    key_obj.write_private_key(desc)
+    private_key_decrypted = desc.getvalue()
+
+    # Store unencrypted private key in the cache
+    current_dict = cache.get("UNENCRYPTED_PRIVATE_KEYS")
+    if current_dict is None:
+        current_dict = dict()
+
+    current_dict[key_pair.id] = private_key_decrypted
+    cache.set("UNENCRYPTED_PRIVATE_KEYS", current_dict)
```

Comment:
I can't find any information on Django cache access time unfortunately, but I imagine this is pretty quick. It looks like Django doesn't have any native support for async caching yet, as per this source and related PR: https://forum.djangoproject.com/t/async-cache-and-a-helper-for-porting-sync-code/302.

That being said, we could still use a celery task to do this independently.

Reply:
May question was related to the whole task. Currently unlocking the key is done in a celery task. My thought was if it quick then it could be just done in the request thread?

Reply:
Moved out of task here: 36d565aae8381eeed7c239bdc9973bfc384fcce3